### PR TITLE
Fn::FindInMap enhancements: default value and additional intrinsic function support

### PR DESCRIPTION
**Pull request by mingxingong**
### Language Enhancement Request For Comment

This is a request for comments about FindInMap enhancements (adding a default value and support for additional intrinsic functions. See #101 for additional details. 

---

## Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+
+We will add support for using other intrinsic functions, including `Fn::Select`, `Fn::Join`, `Fn::Sub`, and `Fn::If` within `Fn::FindInMap`. We will also add an optional default value for `Fn::FindInMap`, which will be the output of the `Fn::FindInMap` if the mapping is not present. The first three parameters of `Fn::FindInMap` will remain the same as before, while the optional fourth parameter `DefaultValue` will be of type `String` or `List`.
+
+# Motivation
+
+## Use of other intrinsic functions:
+
+CloudFormation users may want to use intrinsic functions such as `Fn::Select`, `Fn::Join`, `Fn::Sub`, and `Fn::If` within `Fn::FindInMap`. While CloudFormation users can currently nest `Fn::Ref` and `Fn::FindInMap` in `Fn::FindInMap`, more function support has been requested to give CloudFormation users more flexibility when using `Fn::FindInMap`.
```
Comment (benbridts) [inline RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md:16]: ~~I think we should be explicit about which functions are supported here. Currently one of the example uses `Fn::Split`, but that one is not in this list.~~ This is made explicit at the end

A note about current and future functions provided by the transform could be useful too

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+Mappings:
+  DNS:
+    us:
+      dns: mypage.us.com
+      ttl: '600'
+    ca:
+      dns: mypage.us.com
+      ttl: '600'
+
+!FindInMap [DNS, !Ref country, dns, mypage.default.com]
```
Comment (benbridts) [inline RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md:90]: Is the plan to both support the `MappingPresent` function and a new `!FindInMap` function signature? 

From the details section it seem like only the `!FindInMap` way will be included (which is fine)

~~In either case, I think a definition/explanation of `!FindInMap [Mapping, TopLevel, SecondLevel, Default]` would be useful.~~ This is included at the end

Another way to solve the same problem - with different trade-offs - is to have a special key in the mapping itself:

```
Mappings:
  DNS:
    us:
      dns: mypage.us.com
      ttl: '600'
    ca:
      dns: mypage.us.com
      ttl: '600'
    $default:
      dns: mypage.us.com
      ttl: '600'

# This automatically gets the default, so it's less explicit about what's happening
!FindInMap [DNS, !Ref country, dns]

# It could be made more explicit if needed (could be the only supported way too)
# This option could support multiple user-provided $-keys to get the best from both worlds
!FindInMap [DNS, !Ref country, dns, $default]  

```

For the described use case I think it's more useful to have only one place to define the default, otherwise, if you need to get the value more than one,  the more maintainable way of writing it would be something like

```
Mappings:
  DNS:
    us:
      dns: mypage.us.com
      ttl: '600'
    ca:
      dns: mypage.us.com
      ttl: '600'
  Defaults:
    DNS:
      dns: mypage.default.com
      ttl: 300

!FindInMap [DNS, !Ref country, dns, !FindInMap [Defaults, DNS, dns]]
!FindInMap [DNS, !Ref country, ttl, !FindInMap [Defaults, DNS, ttl]]
```
or
```
Mappings:
  DNS:
    us:
      dns: mypage.us.com
      ttl: '600'
    ca:
      dns: mypage.us.com
      ttl: '600'
    default:
      dns: mypage.default.com
      ttl: 300

!FindInMap [DNS, !Ref country, dns, !FindInMap [DNS, default, dns]]
!FindInMap [DNS, !Ref country, ttl, !FindInMap [DNS, default, ttl]]
```

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+      - HMAC_384
+      - HMAC_512
+      - RSA_2048
+      - RSA_3072
+      - RSA_4096
+      - SYMMETRIC_DEFAULT
+    Default: SYMMETRIC_DEFAULT
+
+Conditions:
+  IsKeyAsymmetricRSA: !Equals ['RSA', !Select [0, !Split [_, !Ref KeySpec]]]
```
Comment (benbridts) [inline RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md:150]: Sidenote: `!StartsWith` or `!MatchRegex` would be useful here (I couldn't immediately find the related issue, but I seem to remember something existing)

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+        "InstanceType": "t2.small"
+      }
+    }
+  }
+}
+```
+
+# Limitations
+
+Users will not be able to nest `Fn::Split` within `Fn::FindInMap` to specify all parameters of `Fn::FindInMap`. Three or four parameters are required in `Fn::FindInMap`.
```
Comment (benbridts) [inline RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md:525]: The example is clear, but I first read this as "you can't use any `Fn::Split`. Is there a way to reword this to make it clear that this applies when you use `Fn::Split` to build the argument list?

Adding a counter example might be useful too, eg:

```yaml
# This is allowed
!FindInMap
  - !Select [0, !Split [ ".", "RegionMap.us-east-1.HVM64" ] ]
  - !Select [1, !Split [ ".", "RegionMap.us-east-1.HVM64" ] ]
  - !Select [2, !Split [ ".", "RegionMap.us-east-1.HVM64" ] ]
```

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+```yaml
+!FindInMap [ !Split [ ".", "RegionMap.us-east-1.HVM64" ] ]
+```
+
+### JSON
+```json
+"Fn::FindInMap": [ { "Fn::Split": [ ".", "RegionMap.us-east-1.HVM64" ] } ]
+```
+
+If users use a reference to a resource property as a key in `Fn::FindInMap`, the `DefaultValue` will be returned, provided that the `DefaultValue` is supplied and does not reference a resource property.
```
Comment (benbridts) [inline RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md:538]: This seems confusing, wouldn't it be better to not allow `!Ref Resource`, `!GetAtt` or `!Sub "${Resource}-${Resource.Att}-...."` in `!FindInMap` ?

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md (hunk 1 above)
Comment (lejiati) [reply RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md:16]: Good point, I think we should be more explicit about what functions are supported under what conditions. Maybe a dedicated section/table listing out all of them(functions and some exceptions such as the `Fn::Select` one) would help?

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md (hunk 2 above)
Comment (lejiati) [reply RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md:90]: That's an interesting idea. 
```yaml
Mappings:
  DNS:
    us:
      dns: mypage.us.com
      ttl: '600'
    ca:
      dns: mypage.us.com
      ttl: '600'
  Defaults:
    DNS:
      dns: mypage.default.com
      ttl: 300
```

Given the fact we would like to support more intrinsic functions in `!FindInMap`, 
```yaml
!FindInMap [DNS, !Ref country, dns, !FindInMap [Defaults, DNS, dns]]
!FindInMap [DNS, !Ref country, ttl, !FindInMap [Defaults, DNS, ttl]]
```
Should be a valid use case.

As for setting a special key in mapping as default key - I'm a bit concerned as it may break customers who already use it as a regular key.

Initially during our internal discussion, we thought adding an optional parameter makes it easy for customers to onboard this feature - you don't need to refactor your `FindInMap` function, just append the argument list then you are good to go. But now I'm thinking, maybe `MappingPresent` can be used without the context of `FindInMap` default value: it's a conditional function so it can alter the behavior of CFN engine, e.g provision a resource conditionally. Yet I don't have any concrete use cases on top of my head to support how useful this new `MappingPresent` function would be.

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md (hunk 4 above)
Comment (lejiati) [reply RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md:525]: Good point! We will reword them and add both valid and invalid examples.

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md (hunk 2 above)
Comment (benbridts) [reply RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md:90]: I didn't check, but I seem to remember some values being not allowed in mapping keys, so depending on the list of values there can be a way to introduce special keys without breaking existing templates

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md (hunk 3 above)
Comment (benbridts) [reply RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md:150]: https://github.com/aws-cloudformation/cfn-language-discussion/issues/87

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md (hunk 2 above)
Comment (lejiati) [reply RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md]: Hi Ben, sorry for the late reply. 

> Another way to solve the same problem - with different trade-offs - is to have a special key in the mapping itself

So I did some investigation on reserving a special key as default. But I decided to rule it out because:
1. IMHO, a reserved key is less flexible compare to adding an optional `DefaultValue` parameter. Customers sometimes may need to define different default values for the same mapping under different circumstances. With `$default` key defined in mapping section, they are stuck with the pre-defined value.
2. It’s hard to define what exactly `$default` would contain: should it be a TopLevelKey or SecondLevelKey? To maintain extensibility, define $default as a SecondLevelKey without TopLevelKey seems make more sense but it would be confusing to customers because `$default` does not maintain the same structure as other entries.
3. This enhancement is delivered via `AWS::LanguageExtensions`, so the response of our transform has to be a legit CloudFormation template. If we reserve a Key without special characters, we run into risk where customers may have already define that key. If we do reserve a key with special characters, we need to remove `$default` section before returning template. This, imo is counter intuitive and does not provide the best customer experience.

> However in the examples us-east-1 is used too. There is a potential to find out the exact allowed pattern by calling aws cloudformation validate-template, but having the documentation be updated is probably better

This is a great catch! *-* is allowed in key and I'll reach out to doc team to make an update.

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md (hunk 5 above)
Comment (lejiati) [reply RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md]: Good point, I've updated RFC to make that clear.

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+If the mapping can be found, it returns the value of the mapping as before, while the `DefaultValue` is returned if the mapping does not exist. Backwards compatibility is maintained by making the `DefaultValue` optional. CloudFormation users will not be required to use other intrinsic functions in `Fn::FindInMap`, they can simply use Strings as they did before.
+
+## Example
+
+### Use of other intrinsic functions in `Fn::FindInMap`
+
+Here is an example using `Fn::Select` and `Fn::Split` in `Fn::FindInMap` - [credit](https://github.com/aws-cloudformation/cfn-language-discussion/issues/73)
+
+```
+Transform: AWS::LanguageExtensions
```
Comment (jlhood) [inline RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md:121]: Should have `AWSTemplateFormatVersion: 2010-09-09` at the top.

File: RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+```
+
+1. `Fn::FindInMap` enhancement is made available via [`AWS::LanguageExtensions` transform](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/transform-aws-languageextensions.html), hence `Fn::FindInMap` can only support operations whose return values are known during transform time. When unsupported intrinsic functions are used, an exception will be returned back to customers.
+    1. Here are examples not supported:
+        1. `Ref` to a resource.
+        2. `Fn::GetAtt` to retrieve an attribute from a resource.
+        3. `Fn::Sub` with nested `Fn::GetAtt` or `Ref` which points to a resource or resource attribute. 
+
+```
+Mapping:
```
Comment (jlhood) [inline RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md:270]: Add a note above this example saying something like. "For example, the following will NOT be supported:" and then add a comment in the YAML just above each `!GetAtt LoadBalancer.DNSName` line saying "GetAtt of a resource attribute not supported by DefaultValue" to make it super clear that this is an example of what will NOT be supported.

**Review by jlhood — CHANGES_REQUESTED** [review]
Just a few minor comments to address but overall looks good!

**Review by jlhood — APPROVED** [review]

**Review by lejiati — APPROVED** [review]
